<html>
	<head>
		<title>Gemma overview</title>
	</head>
	<body>
		<div style="width: 650px">
			<h1>
				Gemma overview
			</h1>

			<p class="emphasized">
				Gemma is an open-source database and software system for the meta-analysis of gene expression data. Gemma contains
				data from hundreds of
				<a href="/Gemma/expressionExperiment/showAllExpressionExperiments.html">public microarray data sets</a>, referencing
				hundreds of
				<a href="/Gemma/bibRef/showAllEeBibRefs.html">published papers</a>.
			</p>


			<p class="emphasized">
				The goals of Gemma are to:
			</p>

			<ul class="plainList">
				<li>
					Aggregate public gene expression studies for the purposes of reanalysis. Our current focus is on human, mouse and
					rat nervous system data.
				</li>
				<li>
					Perform analyses on the expression data and store the results for rapid and easy access.
				</li>
				<li>
					Reanalyze probe sequences to allow consistent cross-platform comparisons.
				</li>
				<li>
					To serve as a framework for the development and evaluation of novel algorithms for expression analysis and
					meta-analysis.
				</li>
				<li>
					Support a range of analytical approaches including but not limited to coexpression and differential expression.
				</li>
				<li>
					Support a variety of expression technologies, including Affymetrix, Illumina and other oligonucleotide arrays, one
					channel and ratiometric cDNA arrays, and (planned) SAGE or other sequence-based techniques.
				<li>
					Provide consistent annotation/descriptions of data sets to enhance usability of the data.
				</li>
				<li>
					Provide external sites with processed data, web services, software and other means of including Gemma tools and
					data in other systems.
				</li>
				<li>
					Allow users to upload expression data for inclusion in Gemma, either "privately" for their own analyses or to be
					made public. (This is a feature under development!)
				</li>
				<li>
					Provide a software system that can be installed and used by others.
				</li>
			</ul>

			<p class="emphasized">
				To these ends, Gemma consists of:
			</p>

			<ul class="plainList">
				<li>
					A database and software framework for storing and accessing expression data with a focus on
					<em>analysis</em>. The software is released under the Apache 2.0 license, which is liberal enough to allow
					commercialization, modification, and reuse by anybody.
				</li>
				<li>
					A set of administrative and curation tools for loading, tracking, processing, annotating, analysis and quality
					control of expression data.
				</li>
				<li>
					A sequence analysis pipeline for probe sequences.
				</li>
				<li>
					Implementations of algorithms for analysis of expression data, in some cases using integration with "R"
				</li>
			</ul>

			<h2>
				The main functionality of Gemma
			</h2>

			<p>
				For users of the web site, Gemma has two main functions, which are supported by various searching and browsing
				functionality.
			</p>

			<h3>
				Coexpression meta-analysis
			</h3>
			<p>
				The entry point for coexpression analyis is
				<a href="/Gemma/searchCoexpression.html">here</a>. The methods used for this "recurring coexpression" analysis
				are essentially as described in
				<a href="http://www.ncbi.nlm.nih.gov/pubmed/15173114">Lee et al, 2004</a>. Briefly, each data set is preprocessed
				and analyzed to identify pairs of genes that are strongly coexpressed. The on-line analysis searches these results
				for pairs of genes that are coexpressed in multiple data sets, starting from a gene you choose.
			</p>

			<h3>
				Differential expression analysis access
			</h3>
			<p>
				We are performing differential expression analysis on most of the data sets in Gemma. Currently these results are
				available from a
				<a href="/Gemma/diff/diffExpressionSearch.html">search form</a> or integrated into the
				<a href="/Gemma/searchCoexpression.html" coexpression analysis</a> results.
			</p>
			<p>
				Data sets are considered for differential expression analysis if they have a factorial design with up to two levels,
				and appropriate replication and block-completeness to allow an anlysis to proceed. Data sets are analyzed using
				t-tests, one-way ANOVA, or two-way ANOVA with or without interaction, as appropriate to the design.
			</p>
			<h2>
				For more information
			</h2>

			<ul class="plainList">
				<li>
					<a href="http://bioinformatics.ubc.ca/Gemma/resources/thanks.html">Credits</a>
				</li>
				<li>
					<a href="http://bioinformatics.ubc.ca/Gemma/resources/faq.html">FAQ</a>
				</li>
				<li>
					<a href="http://bioinformatics.ubc.ca/Gemma/resources/downloads.html">Downloads</a> of software and data.
				</li>
				<li>
					<a href="http://bioinformatics.ubc.ca/Gemma/resources">Resources for developers</a>, including technical
					information about Gemma.
				</li>
			</ul>
		</div>
	</body>
</html>