<body>
<br/>

<p>
	The
	<strong>GO Overlap %</strong> for a pair of co-expressed genes is a
	measure of similarity of the genes, based on prior knowledge. Here it
	is computed as the percentage of overlap between the Gene Ontology
	terms (including all parents) for that Gene pair (queried and
	co-expressed gene).
</p>
<p>
	The percentage is calculated by dividing the number of common GO terms
	by the total number of unique GO terms (including parents) that the
	queried gene has. Values might not be computed for all (or any) of your
	results, in which case a "-" is shown.
</p>

<br/>

<p>
	The <strong>support</strong> for a pair of coexpressed genes is the number of distinct 
	datasets that the gene pair is coexpressed in.
</p>
<ul>
	<li>
		<strong>Positive Correlation</strong> A green number indicates a positive correlation
		between the two genes.
	</li>
	<li>
		<strong>Negative Correlation: </strong> A red number indicates a negative correlation 
		between the two genes.
	</li>
	<li>
		<strong> Adjusted  support</strong> The italicized  number in brackets  is the adjusted  support.
		As probes on an array are non-specific (can hybridize to multiple genes), the adjusted  support 
		only counts datasets that contain at least one specific probe (1-1 mapping between probe and gene)
		for the coexpressed gene in question.  Therefore the adjusted  support helps rule out possible false 
		positives.
	</li>
	
</ul>

<br/>

<p>
	The images shown in this column are a visual representation of all the
	data sets supporting the coexpression results. Each experiment that was
	queried is representd by either a black line or a gap. A black line is
	shown if that experiment contributes to the support for the
	coexpression for the gene pair. A gap is shown if it does not.
</p>
<p>
	The graphic for each gene is shown in the same expression experiment
	ordering, so you can look along vertical 'tracks' to see which
	experiments contribute to many of the links.
</p>

<br />

<p>
	The <strong> sequence analysis report </strong> summarizes our gene assignment analysis
	for the current platform. This summary is broken down into categories
	described below. Details of the analysis procedure are given elsewhere.
</p>
<ul>
	<li>
		<strong>Probes:</strong>The number of elements on the array that form
		the primary analytical unit. This number aggregates features that
		belong to so-called "composite sequences", so on Affymetrix array
		designs this corresponds to "probe sets".
	</li>
	<li>
		<strong> Probes with sequences: </strong> The number of probes that
		have a provided sequence. For some array designs, sequences are
		missing for some or all of its probes. Probes without sequences are
		excluded from further analysis.
	</li>
	<li>
		<strong> Probes with genome alignments: </strong> The number of probes
		whose sequences provide an alignment to the genome. Some probe-genome
		alignments do not necessarily meet stringency requirements and are not
		included in the database.
	</li>
	<li>
		<strong> Probes mapping to genes: </strong> Of probes with alignments,
		how many are mapped to known genes, predicted genes, and probe-aligned
		regions (a breakdown is provided). "Known genes" include RefSeq or
		other annotated genes. "Predicted gene" based on computational
		analyses. A "probe-aligned region" (PAR) is an 'anonymous' genomic
		region that does not fall into the other categories. Given a
		particular alignment, a predicted gene or PAR is only assigned to a
		probe if it does not map to a known or predicted gene. However, a
		probe can be assigned "genes" from all categories when there are
		multiple good alignments to the genome.
	</li>
	<li>
		<strong> Unique genes represented </strong> The number of distinct
		known genes, predicted genes, and probe-aligned regions represented on
		the array. This number can be higher or lower than the number of
		probes mapping because some probes can map to multiple genes, and some
		genes have more than one probe mapping.
	</li>
</ul>
</body>